unstructured/ingest/connector/pinecone.py

```diff
@@ -0,0 +1,128 @@
+
+
+@dataclass
+class PineconeSessionHandle(BaseSessionHandle):
+    service: "pinecone.Index"  # noqa: F821
+
+
+@dataclass
+class PineconeWriteConfig(WriteConfigSessionHandleMixin, ConfigSessionHandleMixin, WriteConfig):
+    api_key: str
+    index_name: str
+    environment: str
+    batch_size: str
+
+    @DestinationConnectionError.wrap
+    @requires_dependencies(["pinecone"], extras="pinecone")
+    def create_pinecone_object(self, api_key, index_name, environment):
+        import pinecone
+
+        pinecone.init(api_key=api_key, environment=environment)
+        index = pinecone.Index(index_name)
+        logger.debug(f"Connected to index: {pinecone.describe_index(index_name)}")
+        return index
+
+    def create_session_handle(self) -> PineconeSessionHandle:
+        service = self.create_pinecone_object(self.api_key, self.index_name, self.environment)
+        return PineconeSessionHandle(service=service)
+
+    @requires_dependencies(["pinecone"], extras="pinecone")
+    def upsert_batch(self, batch):
+        import pinecone.core.client.exceptions
+
+        index = self.session_handle.service
+        try:
+            response = index.upsert(batch)
+        except pinecone.core.client.exceptions.ApiException as api_error:
+            raise WriteError(f"http error: {api_error}") from api_error
+        logger.debug(f"results: {response}")
+
+
+@dataclass
+class SimplePineconeConfig(BaseConnectorConfig):
+    api_key: str
+    index_name: str
+    environment: str
+
+
+@dataclass
+class PineconeDestinationConnector(BaseDestinationConnector):
+    write_config: PineconeWriteConfig
+    connector_config: SimplePineconeConfig
+
+    def initialize(self):
+        pass
+
+    def check_connection(self):
+        pass
+
+    def write_dict(self, *args, dict_list: t.List[t.Dict[str, t.Any]], **kwargs) -> None:
+        logger.info(
+            f"Inserting / updating {len(dict_list)} documents to destination "
+            f"index at {self.connector_config.index_name}",
+        )
+
+        # this is advised to be 100 at maximum in pinecone docs, however when we
+        # chunk content, we hit to the object size limits, so we decrease the batch
+        # size even more here
+        pinecone_batch_size = self.write_config.batch_size
+
+        num_processes = 1
+        if num_processes == 1:
+            for i in range(0, len(dict_list), pinecone_batch_size):
+                self.write_config.upsert_batch(dict_list[i : i + pinecone_batch_size])  # noqa: E203
+
+        else:
+            with mp.Pool(
+                processes=num_processes,
+            ) as pool:
+                pool.map(
+                    self.write_config.upsert_batch,
+                    [
+                        dict_list[i : i + pinecone_batch_size]  # noqa: E203
+                        for i in range(0, len(dict_list), pinecone_batch_size)
+                    ],  # noqa: E203
+                )
+
```

Comment:
I don't think this is going to actually leverage the session as you're intending.

Note: [this logic](https://github.com/Unstructured-IO/unstructured/blob/4b49fbd27f772a02efbc846d9095f2086fedbd11/unstructured/ingest/pipeline/source.py#L9-L25) is what allows the session to persist across a subprocess, by explicitly assigning the session on the doc from the global variable.

Reply:
have you actually tested this? make sure that num_processes > number of batches and confirm that only num_processes number of sessions are created (can log from the create session method)

Reply:
I think I somehow tested it in a wrong way before. 

I thought it was working but after your comment I re-tested and saw that it's not working.

Now, after the commit [fix session handles](https://github.com/Unstructured-IO/unstructured/pull/1774/commits/007ad367942b5efd17f396d2edbfc77fed1cf9c5) : 
- when `number of batches > num_processes`, only `num_processes` sessions get created
- when `num_processes > number of batches`, only `number of batches` sessions get created

These test results fit to my understanding of how session handles work now. Please let me know if there are any concerns on the results